server/src/main/java/org/cloudfoundry/identity/uaa/security/web/CorsFilter.java

```diff
@@ -142,15 +125,66 @@ protected void doFilterInternal(final HttpServletRequest request, final HttpServ
             logger.debug("CORS Processing request: "+getRequestInfo(request));
         }
         if (isXhrRequest(request)) {
-            handleRequest(request, response, filterChain, getXhrConfiguration());
+            handleRequest(request, response, filterChain, resolveXhrCorsConfiguration());
```

Comment:
It looks like this will be called on every request that passes this filter. `resolveXhrCorsConfiguration` doesn't appear to be thread-safe (since it [clears](https://github.com/cloudfoundry/uaa/pull/806/files#diff-5b4ad26806da832e039024c9173adbaeR142) the uri patterns and then [adds valid regex](https://github.com/cloudfoundry/uaa/pull/806/files#diff-5b4ad26806da832e039024c9173adbaeR144) patterns.

I *think* this should be calling a different method to only return the identity zone cors policy (if it has one, otherwise use the 'global' one). This should suffice since regex patterns have already been compiled during the 'initialize' phase.

I think that the `PUT` and `POST` identity zone endpoints should [validate](https://github.com/cloudfoundry/uaa/blob/c1a10a915229e02b6e4f0e75f86efbb9271ed5f7/server/src/main/java/org/cloudfoundry/identity/uaa/zone/GeneralIdentityZoneConfigurationValidator.java#L28) that the cors policy is a valid regex pattern. If we do that we can assume that the cors policy from an identity zone will be valid (even if its created/updated during runtime) removing the need to compile the uri regex patterns on every request.

WDYT?

Reply:
@DennisDenuto, sorry for the late response. The initialize call on this bean only occurs on startup of UAA. This means that we can't compile and store patterns which are configured per zone as that occurs after startup. Unless you mean go through all existing identity zones with cors policies at startup and compile them and store them in a zonified cache? However, the Pattern.compile call is only an optimization. The matcher call will always compile before it does the matching if Pattern.compile has not been called. I like the idea of validating the regex patterns at creation time even though  we would still need to recreate the compiled pattern object each request (unless using a zonified cache). Will update the PR pending your response.

Reply:
Hi @Bitty9 thanks for the reply. I suppose its my turn to apologize for the late reply (this must have gotten lost when i went on vacation). Anyhoo, my concern actually is *not* on the `initialize` method. However it is for the `doFilterInternal` method. 

Specifically, if two identical requests are made in parallel, and if both of these requests reach this filter at the same time, then it seems to me that there is a data race condition problem here. 

i.e. 

Time | Req 1 | Req 2 | notes
------------ | ------------- | ------------- | -------------
0 | reaches filter | reaches filter | 
1 | clears out the patterns zoneXhrConfiguration.getAllowedUriPatterns().clear(); | |
2 | compileCorsAllowedPatterns | |
3 | | clears out the patterns zoneXhrConfiguration.getAllowedUriPatterns().clear();  |
4 | handleRequest with cors filter (refer to notes column)  | | Problem is here because I think the cors filter is now empty because of req 2 
5 | | compileCorsAllowedPatterns |
6 | | handleRequest with cors filter |

Also a reminder that there are also 2 failing tests:
```
org.cloudfoundry.identity.uaa.mock.DefaultConfigurationTestSuite > org.cloudfoundry.identity.uaa.mock.limited.LimitedModeLoginMockMvcTests.testXhrCorsPolicyforNonDefaultZone FAILED
    java.lang.AssertionError: Status expected:<200> but was:<403>
org.cloudfoundry.identity.uaa.mock.DefaultConfigurationTestSuite > org.cloudfoundry.identity.uaa.login.LoginMockMvcTests.testXhrCorsPolicyforNonDefaultZone FAILED
    java.lang.AssertionError: Status expected:<200> but was:<403>
```

Are you able to reproduce these failing tests locally on your machine?